File: LU.py
import scipy
import scipy.linalg # SciPy Linear Algebra Library to compare time of executions ---> our algorithm vs scipy
import numpy as np
import pandas as pd #to upload the data on dataFrame and save it as csv
import time #to measure time of executions
import sys 

def validate(A):
     
     if len(A)!=len(A[0]):
        print("matrix is not square")
        return False
     for i in range(len(A)):
        if A[i,i]==0:
            print("matrix has zero in the main diagonal")
            return False

def LUdecomposition_Optimized(A):
    if validate(A)==False:
        sys.exit(0)
    #initialize n with the number of rows / columns
    n = A.shape[0]
    #initialize matrix U size: ---> nxn with all zeros ---> the zero matrix
    U = np.zeros((n, n), dtype=np.double)
    # initialize matrix L size: ---> nxn to be the identity matrix ---> ones in the main diagonal
    L = np.eye(n, dtype=np.double)
    
    for k in range(n):
        
        U[k, k:] = A[k, k:] - np.matmul(L[k,:k] , U[:k,k:])
        L[(k+1):,k] = (A[(k+1):,k] - np.matmul(L[(k+1):,:] , U[:,k])) / U[k, k]
       
    
    return L, U






def luDecomposition_NonOptimized(A):
    
    if validate(A)==False:
       sys.exit(0)
    #initialize n with the number of rows / columns
    n = A.shape[0]
    
    #initialize matrices lower ,upper size nxn

    lower = [[0 for x in range(n)]
                 for y in range(n)]
    upper = [[0 for x in range(n)]
                 for y in range(n)]
    
   # Decomposing matrix into Upper
    # and Lower triangular matrix
    for i in range(n):
 
        # Upper Triangular
        for k in range(i, n):
 
            # Summation of L(i, j) * U(j, k)
            sum = 0.0
            for j in range(i):
                sum += (lower[i][j] * upper[j][k])
 
            # Evaluating U(i, k)
            upper[i][k] = A[i][k] - sum
 
        # Lower Triangular
        for k in range(i, n):
            if (i == k):
                lower[i][i] = 1  # Diagonal as 1
            else:
 
                # Summation of L(k, j) * U(j, i)
                sum = 0.0
                for j in range(i):
                    sum += (lower[k][j] * upper[j][i])
 
                # Evaluating L(k, i)
                lower[k][i] = float((A[k][i] - sum) /
                                  upper[i][i])
    return lower,upper





 



    """
 ____________  main-test _____________________________
"""

print("start reading file...") 

fileName = "data_10.txt"

input_matrix = np.loadtxt(fileName, delimiter=',' , dtype = np.float64)



print("data fetched successfully...")
print("number of lines in file:", input_matrix.shape[0])
while(True):
     choice = int(input("Enter which version do you want to use? \n 0 - non-optimized \n 1 - optimized\n"))
     if choice == 0 or choice == 1:
         break 
        
     else:
         print("\n invalid option please enter again...")

print("LU decomposition start...")
#measuring time using the python library
print("starting scipy python LU decomposition...")
t0 = time.time()
scipy.linalg.lu(input_matrix) # --> python scipy LU decomposition
t1 = time.time() - t0


print("python scipy LU decomposition time:",t1)
#measuring time using our LU decomposition code
print("starting OUR LU decomposition...")
t2 = time.time()

if choice == 1:
   lower , upper = LUdecomposition_Optimized(input_matrix) 
elif choice == 0:
       lower , upper = luDecomposition_NonOptimized(input_matrix) 
t3 = time.time() - t2

print("Our LU decomposition time:",t3)
#to check if input matrix is the same as LU
res = np.matmul(lower,upper);
for i in range(res.shape[0]):
  for j in range(res.shape[0]):
        if (res[i][j] != input_matrix[i][j]):
         isIdentical=0
        else:
         isIdentical=1

if(isIdentical == 1):
    print("matrix identical")
else:
    print("matrix not identical")
    
#to check if input matrix is the same as LU
print(np.allclose(input_matrix, np.matmul(lower,upper)))

print("saving to file...") 
t4 = time.time()
print("saving to file lower...") 
DFL = pd.DataFrame(lower)
DFL.to_csv("L.csv")
print("saving to file upper...") 

DFU = pd.DataFrame(upper)
DFU.to_csv("U.csv")

arrL = np.array(lower)
arrU = np.array(upper)

np.savetxt('L.txt',arrL, delimiter=',',fmt="%.4f")
np.savetxt('U.txt',arrU, delimiter=',',fmt="%.4f")

t5 = time.time() - t4
print("time to save to files: ", t5)

File: LU.cpp
#include <iostream>
#include <fstream> // For using files
#include <string>  //for using string operations
#include <vector>  //for vector temp
#include <chrono> // For measuring time of executions

using namespace std;

/**Program to calculate LU decomposition of a square matrix

 Input:Square matrix
 Output:4 files L.txt , L.csv and U.txt , U.csv containing: Lower triangular matrix L , Upper triangular matrix U

**/



//Functions prototypes
void initMatrices(double**& L, double**& U, int n);
void LUdecomposition(double**& A, double**& L, double**& U, int n);
void saveResults(double** matrix, int n, string fileName);
double** readFile(int& n, string fileName);


//Function to initialize matrices: L:lower , U:upper ---> size:nxn
void initMatrices(double**& L, double**& U, int n) {

    L = new double* [n];
    U = new double* [n];
    for (int i = 0; i < n; i++)
    {
        L[i] = new double[n];
        U[i] = new double[n];
    }
}

//LU decomposition fucntion
void LUdecomposition(double**& A, double**& L, double**& U, int n) {

    double sum;
    //first start by initializing matrices L:lower , U:upper --->size:nxn 
    initMatrices(L, U, n);

    for (int i = 0; i < n; i++)
    {
        //Calculation of Upper Triangular matrix
        for (int k = i; k < n; k++)
        {
            // Calculating the Sum of L(i, j) * U(j, k)
            sum = 0.0;
            for (int j = 0; j < i; j++)
                sum = sum + (L[i][j] * U[j][k]);

            //Calculating U(i, k)
            U[i][k] = A[i][k] - sum;

            //Putting zeros in the lower triangular
            if (i < k) {
                U[k][i] = 0;
            }
        }

        //Calculation of lower Triangular matrix
        for (int k = i; k < n; k++)
        {
            if (i == k)
                L[i][i] = 1; //Putting ones in the main diagonal
            else
            {
                //Calculating the Sum of L(k, j) * U(j, i)
                sum = 0.0;
                for (int j = 0; j < i; j++)
                    sum = sum + (L[k][j] * U[j][i]);

                //Calculating L(k, i)
                L[k][i] = (A[k][i] - sum) / U[i][i];
            }
            //Putting zeros in the upper triangular
            if (k > i) {
                L[i][k] = 0;
            }
        }
    }
}

//Function to read matrix from txt file
double** readFile(int& n, string fileName) {


    string line;
    string delimiter = ",";
    size_t pos = 0;
    string num_str;

    vector<double> temp;
    // Open the file to read
    ifstream dataFile(fileName);
    //if fail to read
    if (dataFile.fail()) {
        cerr << "Error opening file: " << fileName << endl;
        exit(1);
    }

    cout << "Reading the data from file: " << fileName << endl;

    //get data from file line by line and count the number of rows.
    while (getline(dataFile, line))
    {
        //number of rows in file
        n++;


        while ((pos = line.find(delimiter)) != string::npos)// --> get numbers from line
        {

            num_str = line.substr(0, pos);//-->number in line

            temp.push_back(stod(num_str));//-->push back the number in line to vector
            line.erase(0, pos + delimiter.length());

        }

        temp.push_back(stod(line));
    }
    cout << "Finished reading the data from file: " << fileName << endl;

    cout << "Number of lines read:" << n << endl;

    //Create matrix and fill it with data from file
    double** matrix = new double* [n];
    for (int i = 0; i < n; i++) {
        matrix[i] = new double[n];
    }

    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {

            matrix[i][j] = temp[j + i * n];


        }
    }

    return matrix;
}

//Function to save results in file
void saveResults(double** matrix, int n, string fileName) {

    cout << "Saving result in: " << fileName << endl;
    ofstream output(fileName);
    for (int i = 0; i < n; i++)
    {
        for (int j = 0; j < n; j++)
        {
            //if not the end of the line write the number to file with -->,
            if (j != n - 1) {
                output << matrix[i][j] << ",";
            }
            //make sure not to write to file -->, at the end of the line
            if (j == n - 1) {
                output << matrix[i][j];
            }
        }
        //write new line at the end of the current line
        output << "\n";
    }
}



int main() {

    //initializations
    int n = 0; //Matrix size
    double** A; //Matrix to decompose
    double** L; //Lower Triangular matrix
    double** U; //Upper Triangular matrix
    string fileName = "data_10.txt";

    //Measuring execution time of reading matrix from txt file
    auto start0 = chrono::high_resolution_clock::now();
    A = readFile(n, fileName); //Reading matrix from txt file
    auto elapsed0 = chrono::high_resolution_clock::now() - start0;//how much time elapsed from start0
    auto secondsTimeRead = std::chrono::duration_cast<std::chrono::seconds>(elapsed0).count();
    cout << "time readFile: " << secondsTimeRead << "s" << endl;



    //Measuring execution time of LU decomposition 
    auto start1 = chrono::high_resolution_clock::now();
    cout << "Starting LU Decomposition..." << endl;
    LUdecomposition(A, L, U, n);
    auto elapsed1 = chrono::high_resolution_clock::now() - start1;
    auto secondsLU = std::chrono::duration_cast<std::chrono::seconds>(elapsed1).count();
    cout << "time LU decomposition: " << secondsLU << "s" << endl;



    //Measuring execution time of saving results to file
    auto start2 = chrono::high_resolution_clock::now();
    saveResults(L, n, "L.txt");
    saveResults(U, n, "U.txt");
    saveResults(L, n, "L.csv");
    saveResults(U, n, "U.csv");
    auto elapsed2 = chrono::high_resolution_clock::now() - start2;
    auto secondsSaveResults = std::chrono::duration_cast<std::chrono::seconds>(elapsed2).count();
    cout << "time saveResults: " << secondsSaveResults << "s" << endl;


    cout << "Finished" << endl;



    char ch = getchar();
    return 0;
}


